uswds/uswds
# USWDS - Table: Add sticky headers

# Summary

**New table variant.** Created the `.usa-table--sticky-header` variant to allow sticky positioning on table headers. 

> [!CAUTION]
> Sticky headers are not compatible with the scrollable and stacked table variants.

Thanks to @etanb for originating this work in PR https://github.com/uswds/uswds/pull/5074.

## Breaking change

This is not a breaking change.

## Related issue

Closes https://github.com/uswds/uswds/issues/5077

## Related pull requests

[Changelog and documentation PR](https://github.com/uswds/uswds-site/pull/2229)

## Preview link

- [Sticky table](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-sticky-table-header/?path=/story/components-table--sticky-header)
- [Default table](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-sticky-table-header/?path=/story/components-table--default)
- [Borderless table](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-sticky-table-header/?path=/story/components-table--borderless)
- [Striped table](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-sticky-table-header/?path=/story/components-table--striped)
- [Sortable table](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-sticky-table-header/?path=/story/components-table--sortable)

## Problem statement

For very long tables, it can be difficult to remember the associated header content. Creating the option to have table headers stick to the top of the page as the user scrolls would help keep tabular data in context.

## Solution

This change creates a `.usa-table--sticky-header` variant that gives the table header sticky positioning. 

> **Note**
> At this time, sticky headers are not compatible with the scrollable and stacked table variants. For the scrollable variant, this incompatibility is related to a known issue ([on W3C CSS repo](https://github.com/w3c/csswg-drafts/issues/865)) with sticky positioning and scrolling overflow. If we want to pursue a sticky header that is compatible with scrollable tables, I recommend we open a separate issue because it will require a decent lift to find a workaround. 

### New variant

| Variant name              | Description | 
| ---------------------------- | :-------------- |
| `.usa-table--sticky-header` |   Allows the table header element to stick to the top of the page as the user scrolls. Preferred for long tables with many rows. This variant is not compatible with the scrollable and stacked table variants. |   

### New settings

| Setting name              | Description | 
| ---------------------------- | :-------------- |
| `$theme-table-background-color` |   Defines the table background color. A value of `default` uses the body background color. |   
| `$theme-table-sticky-top-offset` |  Defines the top offset for sticky headers. |

## Testing and review
- Confirm that sticky positioning works with the default, borderless, striped, and sortable table variants.
- Test on a variety of browsers
- Confirm that sticky headers are compatible with multiple header rows
- Confirm that the new settings work as expected
- Confirm no visual regressions

To test in Storybook:
1. In a variety of browsers, open each table variant story from the list of preview links
1. In each story, turn on the sticky header control 
1. Confirm that the table header sticks to the top of the page when you scroll down
    > **Note** 
    > Sticky headers will not stick in scrollable and stacked variants. If the header is not sticking, confirm that the `scrollable` control is turned off.  
1. Turn on the control for multiple header rows
1. Confirm that both header rows stick to the top of the page when you scroll down and confirm that there are no visual irregularities (like table content showing behind the header on scroll)
